Beta (WIP)

onalant opened the pull request:
Continuation of #6.

Commit:
Starting class-based implementation

Need to read more about ``typing'' module.

Commit:
Still thinking about how to structure table loading.

Otherwise, continuing to bring in isomorphic functionality from
old interface (Ex: pose transformations and chemical information)

Commit:
More functionality

The coordinates-to-frame method can /technically/ be an extension of
the rays-to-frame method, but the implementation doesn't look any
nicer.

ResidueTable can be accessed with ``[]'' by providing a uint64 key.
Method for accessing by ray pair will be added, and will utilize
this access method.

Better names probably needed.

Commit:
Using HDF5 for data format

Table is going to transparently handle the indexing, and the client
will be responsible for handling data from the table.

Commit:
Table loading and slicing ready, need to write client.

This could just be a separate repository, though refactoring will
definitely be needed. The code is kind of nasty.

Commit:
Minor refactoring.

Considering making one contiguous top-level dataset with slices into
relevant sections. Possibly something to consider, though we would
be giving up lazy-indexing by doing this.

Commit:
Very sensible suggestion by Adam.

Commit:
LRU cache option in case memory is an issue.

May or may not be important.

Commit:
Basic implementation of client, still plenty of loose ends to fix up.

- ``informal_requirements.txt'' is named that way for a reason. Do not
	try to pip install it.
- ``process.py'' is terribly named. Fix.
- ``table.py'' is going to be moved to a separate module.
- ``util.py'' is just a set of method bindings to make working with
	pyrosetta classes easier.

Commit:
I may have blown up the pip installation. We will see...

Commit:
Figured out /some/ RIF exceptions, still confused about linalg errors.

Works, maybe.™

Commit:
Fixed critical error in `table.py`.

Probably left over from earlier debugging.

Commit:
Minor style preferences and gitignore fix.

Commit:
Clash check and minimization

Commit:
Removed LRU and type annotations

Commit:
Primitives of ray selection

Commit:
Shuffling files around

Commit:
Sc_Bb, Sc_ScBb, and Sc_Sc ray selectors with fixed ray order.

Probably need to fix data, if rays were hashed in arbitrary order.
Resolves onalant/privileged_residues#5.

Commit:
Added network selection.

Forgot topology of sidechain ray lists. Fixed.

Side note: Sc_Sc ray selection is currently broken. How do we handle
donor-donor or acceptor-acceptor cases without infecting the
donor-acceptor rays? The rays are not interchangeable!

Commit:
Some fixes and improvements to ray selection.

Commit:
Still trying to get everything working together.

Writing basic sanity tests.

Commit:
Fixed module imports.

Commit:
Selection of test file.

Commit:
Matching not working.

Commit:
Dumping poses now, but bugged

Commit:
Forgot that atom references were const

weitzner commented on privileged_residues/util.py:
```diff
@@ -18,9 +18,9 @@ def apply_transform(self, xform):
-            atom.xyz(V3(*transformed_coords[:3, index]))
+    for i in range(1, len(self) + 1):
+        for j in range(1, self.residue(i).natoms() + 1):
+            self.residue(i).atom(j).xyz(V3(*transformed_coords[:3, index]))
```
See `pyrosetta.rosetta.core.conformation.Residue.apply_transform_Rx_plus_v`

```
Docstring:
apply_transform_Rx_plus_v(self: pyrosetta.rosetta.core.conformation.Residue, R: pyrosetta.rosetta.numeric.xyzMatrix_double_t, v: pyrosetta.rosetta.numeric.xyzVector_double_t) -> None

Applies a transform of the form Rx + v, where R is a rotation
 matrix, V is a vector, and x is the original position in xyz space

C++: core::conformation::Residue::apply_transform_Rx_plus_v(class numeric::xyzMatrix<double>, class numeric::xyzVector<double>) --> void
Type:      method
```

Since the same transform is being applied to all of the coordinates in the `Pose`, you could do the following:
```python
def xyzMatrixFromNumpy(np_arr):
    rot = pyrosetta.rosetta.numeric.xyzMatrix_double_t(0.)
    rot.xx = np_arr[0, 0]
    rot.xy = np_arr[0, 1]
    rot.xz = np_arr[0, 2]

    rot.yx = np_arr[1, 0]
    rot.yy = np_arr[1, 1]
    rot.yz = np_arr[1, 2]

    rot.zx = np_arr[2, 0]
    rot.zy = np_arr[2, 1]
    rot.zz = np_arr[2, 2]

rot = xyzMatrixFromNumpy(xform[:3, :3])
trans = pyrosetta.rosetta.numeric.xyzVector_double_t(*xform[:3, 3])

for rsd in self.residues:
    rsd.apply_transform_Rx_plus_v(rot, trans)
```

Similar functionality to the `xyzMatrixFromNumpy` function above is provided in the `pyrosetta.toolbox.numpy_utils` module, but it's a little sketchy and could stand some improvement.

So the full function could be:

```python
from pyrosetta.toolbox.numpy_utils import numpy_to_rosetta as xyzMatrixFromNumpy

@bind_method(pyrosetta.rosetta.core.pose.Pose)
def apply_transform(self, xform):
    """description - note the type and shape of the arguments"""

    assert(xform.shape == (4, 4)) # homogeneous transform

    rot = xyzMatrixFromNumpy(xform[:3, :3])
    trans = pyrosetta.rosetta.numeric.xyzVector_double_t(*xform[:3, 3])

    map(lambda rsd: rsd.apply_transform_Rx_plus_v(rot, trans), self.residues)
```

onalant replied:
Was not aware of this method. Thanks for letting me know. ✔️ ✔️

onalant replied:
Resolved by 76d3eb6.

Commit:
Modify implementation of apply_transform to use PyRosetta API.

Suggested by Brian Weitzner in weitzner/privileged_residues#7.

Commit:
Testing works.

Also fixed bug with sidechain ray identification.

Commit:
Fixed some bugs in the ray selection process.

Tests pass, though need to remove randomness and test each interaction
explicitly.

Commit:
Class implementation of tests to see sub-test statuses.

Commit:
More fiddling with tests

Commit:
Installation instructions for test requirements

Commit:
Change that should increase correctness of output

weitzner commented on privileged_residues/geometry.py:
```diff
@@ -0,0 +1,81 @@
+    ray[-1, 0] = 1.  # point!
+
+    ray[:-1, 1] = direction
+    ray[-1, 1] = 0.  # point!
```
comment should be `# direction!`

onalant replied:
Resolved by 4c1dc50

✔️ ✔️

Commit:
Knowing the RMSD for failed tests would be nice

Commit:
A rudimentary test definition for networks, but very slow

Commit:
NumPy-style documentation

weitzner commented on privileged_residues/util.py:
```diff
@@ -0,0 +1,99 @@
+
+    assert(xform.shape == (4, 4)) # homogeneous transform
+
+    M = numpy_to_rosetta(xform[:3, :3])
```
It may make sense to call `M` `Rx` here.

Commit:
Demo script

weitzner commented on privileged_residues/util.py:
```diff
@@ -0,0 +1,99 @@
+        A homogeneous transform.
+    """
+
+    assert(xform.shape == (4, 4)) # homogeneous transform
```
There's a better assertion here. I think rotations have determinant == 1. so let's check Rx for that.

Commit:
Updated basic matcher.

Asserts in geometry.py cause issues since there are some degenerate
cases where the vectors are too short to normalize properly.

Commit:
Postprocessing is actually halfway decent now.

Commit:
More useful assertions.

weitzner commented:
This all all looks great. There are merge conflicts, though. Pull upstream and resolve the conflicts and I will hit the button!

Commit:
Merge branch 'master' into beta

Commit:
Cutoff for matcher.

Commit:
Revert "Cutoff for matcher."

This reverts commit f3fc4f6d84355e9eac88be1053b7edc21dcde091.

Commit:
Better implementation of limited output.

Commit:
Marginally better basic matcher, but probably need to have dedicated
limiter for raw data.

Commit:
Reduced output actually reduces runtime now. Marvelous.

ghost commented:
@weitzner Are there any other blocking issues for the v0.9 merge?

weitzner commented:
Looks good to me. Merging in.
